File: src/main/java/aqua/recommend/RPSimilarAnime.java
package aqua.recommend;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class RPSimilarAnime {
    private static final int ITEM_MIN = 10;
    private static final int ITEM_MAX = 20;

    public final Map<Integer, Integer> animeIndexMap;
    public final int[] animeRatedMap;
    public final int similarAnimeCount;
    public final int[] similarAnimeId;
    public final float[] similarAnimeScore;

    public RPSimilarAnime(Map<Integer, Integer> animeIndexMap, int similarAnimeCount, int[] similarAnimeId, float[] similarAnimeScore) {
        this.animeIndexMap = animeIndexMap;
        this.similarAnimeCount = similarAnimeCount;
        this.similarAnimeId = similarAnimeId;
        this.similarAnimeScore = similarAnimeScore;

        this.animeRatedMap = new int[animeIndexMap.size()];
        for (Map.Entry<Integer, Integer> entry : animeIndexMap.entrySet())
            animeRatedMap[entry.getValue()] = entry.getKey();
    }

    public RPSimilarAnime(int[] animeRatedMap, int similarAnimeCount, int[] similarAnimeId, float[] similarAnimeScore) {
        this.animeRatedMap = animeRatedMap;
        this.similarAnimeCount = similarAnimeCount;
        this.similarAnimeId = similarAnimeId;
        this.similarAnimeScore = similarAnimeScore;

        this.animeIndexMap = new HashMap<>();
        for (int i = 0; i < animeRatedMap.length; ++i)
            animeIndexMap.put(animeRatedMap[i], i);
    }

    public List<ScoredAnimeId> similarAnime(int animedbId) {
        Integer index = animeIndexMap.get(animedbId);
        if (index == null)
            return new ArrayList<>();
        List<ScoredAnimeId> result = new ArrayList<>(similarAnimeCount);

        for (int i = 0, currentIndex = index * similarAnimeCount; i < similarAnimeCount; ++i, ++currentIndex) {
            if (similarAnimeId[currentIndex] == 0)
                break;
            result.add(new ScoredAnimeId(similarAnimeId[currentIndex], similarAnimeScore[currentIndex]));
        }

        return result;
    }

    public List<ScoredAnime> findSimilarAnime(CFUser user) {
        List<ScoredAnimeId> sortedScores = new ArrayList<>(user.completedAndDroppedIds.length);
        for (int i = 0; i < user.completedAndDroppedIds.length; ++i) {
            sortedScores.add(new ScoredAnimeId(user.completedAndDroppedIds[i], user.completedAndDroppedRating[i]));
        }
        sortedScores.sort(ScoredAnimeId.SORT_SCORE);
        int maxUse = Math.min(ITEM_MAX, sortedScores.size());
        int minUse = Math.min(ITEM_MIN, sortedScores.size());
        List<ScoredAnimeId> picked = new ArrayList<>(maxUse);
        for (int i = 0; i < maxUse; ++i) {
            picked.add(sortedScores.get(sortedScores.size() - 1 - i));
        }
        Collections.shuffle(picked);
        int use = (int) (Math.random() * (maxUse + 1 - minUse)) + minUse;
        Map<Integer, ScoredAnimeId> recommendations = new HashMap<>();
        for (int i = 0; i < use; ++i) {
            ScoredAnimeId likedItem = picked.get(i);
            Integer likedAnimeIndex = animeIndexMap.get(likedItem.animedbId);
            if (likedAnimeIndex == null)
                continue;
            for (int j = 0, index = likedAnimeIndex * similarAnimeCount; j < similarAnimeCount; ++j, ++index) {
                int animedbId = similarAnimeId[index];
                if (animedbId == 0)
                    continue;
                float score = similarAnimeScore[index] * likedItem.score;
                ScoredAnimeId recommendation = recommendations.computeIfAbsent(animedbId, id -> new ScoredAnimeId(id, 0.0f));
                recommendation.score += score;
            }
        }
        return new ArrayList<>(recommendations.values());
    }
}

File: src/main/clojure/aqua/recommend/rp_similar_anime.clj
(ns aqua.recommend.rp-similar-anime
  (:require aqua.recommend.item-item-model))

(defn create-rp-similarity [user-list anime-map rank similar-count]
  (let [anime-index-map (java.util.HashMap.)
        anime-rated-map (java.util.ArrayList.)]
    (doseq [user user-list]
      (doseq [rated (.animeList user)]
        (.putIfAbsent anime-index-map
                      (.animedbId rated)
                      (.size anime-index-map))))
    (dotimes [_ (.size anime-index-map)]
      (.add anime-rated-map nil))
    (doseq [entry anime-index-map]
      (.set anime-rated-map (.getValue entry) (.getKey entry)))
    (let [compute (aqua.recommend.ComputeRPSimilarAnime. anime-map anime-index-map similar-count)]
      (.findSimilarAnime compute user-list rank)
      (.rpSimilarAnime compute))))

(defn similar-anime-debug [rp-similarity anime-id anime-map]
  (let [similar-scored (.similarAnime rp-similarity anime-id)]
    (doseq [scored similar-scored]
      (let [anime (anime-map (.animedbId scored))]
        (println (.animedbId scored) (.title anime) (.score scored))))))

(defn store-rp-similarity [out rp-similarity]
  (aqua.recommend.item-item-model/store-item-item out rp-similarity))

(defn load-rp-similarity [path]
  (aqua.recommend.item-item-model/load-item-item path))

(defn get-recommendations [user
                           ^aqua.recommend.ItemItemModel rp
                           remove-known-anime]
  (aqua.recommend.item-item-model/get-recommendations user rp remove-known-anime))

(defn get-all-recommendations [user rp remove-known-anime keep-airing-anime tagger]
  (aqua.recommend.item-item-model/get-all-recommendations user rp remove-known-anime keep-airing-anime tagger))
